UART is now stable.

PyroAVR opened the pull request:
+ Documentation updated in files and headers
 + Submodules added for queue library

 - Removed stale code and files from previous asm driver
 - Removed unused kernel interfaces sbrk, abort
 - Renamed kernel_support to file_table

PyroAVR committed:
UART is now stable.

 + Documentation updated in files and headers
 + Submodules added for queue library

 - Removed stale code and files from previous asm driver
 - Removed unused kernel interfaces sbrk, abort
 - Renamed kernel_support to file_table

spressomartini commented on src/drivers/kinetis/uart.c:
```diff
@@ -0,0 +1,278 @@
+    {UART2, &txq_headers[2], &rxq_headers[2]},
+    {UART3, &txq_headers[3], &rxq_headers[3]},
+    {UART4, &txq_headers[4], &rxq_headers[4]},
+    {UART5, &txq_headers[5], &rxq_headers[5]}
```
we don't have access to UART5

PyroAVR replied:
generic driver for K64

spressomartini commented on src/drivers/kinetis/uart.c:
```diff
@@ -0,0 +1,278 @@
+        conf.uart_base->RWFIFO = 1;
+        // set TWFIFO[TXWATER] such that interrupts are actually generated
+        /*
+         *conf.uart_base->TWFIFO = 1;
```
stray comments o_o

PyroAVR replied:
these are not stray - this is something that we can optionally enable.  Do you want it enable-able by the config struct?

spressomartini replied:
probably a good idea

PyroAVR replied:
Fixed now - these are configurable per-uart.

spressomartini commented on src/drivers/kinetis/uart.c:
```diff
@@ -0,0 +1,278 @@
+void uart_isr(void);
+
+// currently active uart
+static volatile int which_uart;
```
cant this global be avoided if the UART number is passed into the generic ISR

PyroAVR replied:
Fixed now - constant is passed in at the entry point functions.

spressomartini commented on src/drivers/kinetis/uart.c:
```diff
@@ -0,0 +1,278 @@
+queue_t rxq_headers[6];
+volatile char txqs[6][255];
+volatile char rxqs[6][255];
+uart_context contexts[6] = {
```
does every queue get created even if we don't use a specific UART?

PyroAVR replied:
yes, they have to be statically allocated, as we do not have dynamic memory in the kernel yet.  When we have that, these can go.

spressomartini commented on src/drivers/kinetis/uart.c:
```diff
@@ -0,0 +1,278 @@
+// currently active uart
+static volatile int which_uart;
+// true if any uart is currently being processed by uart_isr
+static volatile bool isr_active;
```
whats going on with this? do we need to lock per iqr_handler?

PyroAVR replied:
probs not, time to find out if it breaks

PyroAVR replied:
Apparently not, I removed this.

PyroAVR committed:
Removed stray code, fixed inclusions

PyroAVR committed:
FIFOs are now configurable

Interrupts are coalesced if FIFOs are enabled, which greatly enhances
system performance

PyroAVR committed:
More stale code removal

PyroAVR commented:
test branch is `test/uart_subsystem`.
Pull from that branch and run on the FRDM board.  The LED will flicker between red and blue if there is significant latency in interrupt handling.  It will appear purple if the system is running smoothly.
Torture test: `$ yes > /dev/k64_serial_port_name`

The system will start to stutter under the torture test without FIFOs enabled, but once they are turned on, the operation returns to normal (no flickering).

spressomartini commented on src/drivers/kinetis/uart.c:
```diff
@@ -0,0 +1,269 @@
+    queue_init(&txq_headers[which], &txqs[which], 255);
+    queue_init(&rxq_headers[which], &rxqs[which], 255);
+
+    // make isr available
```
whats going on with `which_uart` and `isr_active` here?

PyroAVR replied:
e r g, I'll fix it

spressomartini commented on src/drivers/kinetis/uart.c:
```diff
@@ -0,0 +1,269 @@
+
+void uart_isr(int which_uart){
+    uint8_t sz = 0;
+    asm("cpsid i");
```
whats the purpose of turning off interrupts here?

PyroAVR replied:
No interruption during the ISR - if we allow interruption, the data may be lost (this was likely one of the causes of bugs in the asm uart driver)

PyroAVR committed:
Removed stale code for isr context locking

PyroAVR committed:
Removed cpsi instructions, see test/uart_subsystem

PyroAVR committed:
Added base_queue dependency (for real this time)

PyroAVR committed:
Slight performance enhancement on TX IRQ

 + better txfifo words-remaining calculation

PyroAVR committed:
More performance improvements, hammering bug fixed

 + tx polling loop now exits early if no data is available in the sw
 queue
 + tx int. is disabled to prevent self-preemption until the isr exits

PyroAVR committed:
Added optional blocking to UARTs

The appropriate flag is __KINETIS_UART_BLOCKING_ISR.
If defined, the uart isr issues cpsid i at the beginning and cpsie i at
the end of the isr.

spressomartini reviewed (approved):
Looks good to me! 
Remember to make note of the blocking compiler flag, the compile submodules statically flag, and if no blocking is used, use higher interrupt priority for the UART.

PyroAVR commented:
Those changes will be made in a future pr to update documentation, as well as some quality of life updates (like configurable buffers).  Merging for now so that we can move on with development until this is critical again.
